### ISSUE TITLE
 Insufficient color contrast for calendar selected date and Publish button

### ISSUE THREAD

**Issue by karlgroves** [WordPress/gutenberg#15278]
Insufficient color contrast for calendar selected date and Publish button 

-   **Severity:** Medium
-   **Affected Populations:**
    -   Low-Vision
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Document Panel
    -   Publish and Unpublish

#### Issue description
The selected date within the calendar, and the "Publish" button, have
insufficient color contrast, below the minimum threshold of 4.5:1 for
text:

-   "Publish" button: #fff (white) on #0085ba (blue): **4.15:1**
-   Selected date within the calendar: #fff (white) on #0085ba (blue): **4.15:1**

Sufficient color contrast is important for users who have low-vision or
are color-blind, because text with a low contrast ratio may be difficult
or impossible for such users to see.

##### Issue Code

```
    /* selected calendar date */
    .components-datetime__date .CalendarDay__selected {
        background: #0085ba;
    }

    .CalendarDay__selected,
    .CalendarDay__selected:active,
    .CalendarDay__selected:hover {
        ...
        color: #fff;
    }


    /* Publish button */
    .components-button.is-primary {
        background: #0085ba;
         ...
        color: #fff;
    }
```

#### Remediation Guidance
Darken the background color to achieve a minimum contrast of 4.5:1.

##### Recommended Code

```
    /* selected calendar date */
    .components-datetime__date .CalendarDay__selected {
        background: #006a95;
    }

    .CalendarDay__selected,
    .CalendarDay__selected:active,
    .CalendarDay__selected:hover {
        ...
        color: #fff;
    }

    /* Publish button */
    .components-button.is-primary {
        background: #006a95;
         ...
        color: #fff;
    }
```

#### Relevant standards

-   [1.4.3 Contrast (Minimum) (Level
    AA)](https://www.w3.org/TR/WCAG20/#visual-audio-contrast-contrast)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-19 in Tenon's report

**Comment by mapk** [comment]
Thanks. This appears to be the standard WordPress primary button styles across the board, not just Gutenberg.

Here's the conversation on this in Core Trac. https://core.trac.wordpress.org/ticket/31459

Before closing this, I'd like to get Accessibility Team's input on this. @afercia @LukePettway 

**Examples of this same button used elsewhere:**

w.org header
<img width="346" alt="Screen Shot 2019-05-01 at 6 28 17 PM" src="https://user-images.githubusercontent.com/617986/57052905-418cd980-6c3f-11e9-9421-696bafb8d0df.png">

Download page on w.org
<img width="341" alt="Screen Shot 2019-05-01 at 6 28 52 PM" src="https://user-images.githubusercontent.com/617986/57052907-481b5100-6c3f-11e9-89b2-10f8e2e49efb.png">

Login page
<img width="367" alt="Screen Shot 2019-05-01 at 6 29 33 PM" src="https://user-images.githubusercontent.com/617986/57052911-4e113200-6c3f-11e9-8fd6-1ffe26a1687f.png">

**Comment by afercia** [comment]
**Publish button (and all the primary blue buttons in WordPress):**
Yes, they were previously discussed in https://core.trac.wordpress.org/ticket/31459. In order to not touch the color palette, we opted to add a subtle text-shadow around the white text to increase the perceived contrast. See https://core.trac.wordpress.org/changeset/34948

This technique is explicitly mentioned in https://www.w3.org/TR/WCAG20-TECHS/G18.html. However, it's a compromise and the best option would be to just darken the `#0085ba` blue. This part should be reported on the core Trac.

**Selected date within the calendar**
(and all the other controls in Gutenberg that use a `#0085ba` background without the text-shadow technique) Contrast is insufficient. I'd say it's one more good reason to just darken the `#0085ba` blue, as the text-shadow technique can be easily overlooked and can be undesirable on some elements.

**Comment by mapk** [comment]
Just for reference, the selected date within the calendar looks like this:

<img width="274" alt="Screen Shot 2019-05-02 at 9 15 45 AM" src="https://user-images.githubusercontent.com/617986/57090140-f57f7a80-6cba-11e9-9b63-b6aeda4087f5.png">

**Comment by melchoyce** [comment]
The next color up in our [color chart](https://codepen.io/hugobaeta/full/RNOzoV/) that passes contrast is  `#307bad`, which has a contrast ratio of 4.6:

![image](https://user-images.githubusercontent.com/2846578/57098718-96783080-6ccf-11e9-9133-42c9ff0e0f7b.png) 
![image](https://user-images.githubusercontent.com/2846578/57098761-b27bd200-6ccf-11e9-9708-3d24d93fb0f3.png)

That would also allow us to remove that text shadow, which would be nice.

**Comment by mapk** [comment]
Alrighty, as mentioned in the [triage slack chat](https://wordpress.slack.com/archives/C02S78ZAL/p1556822386479600), let's break out the primary button issue into a Core Trac issue and work on the calendar date highlight in Gutenberg.

**Comment by sarahmonster** [comment]
Note: currently that #307bad isn't available as a colour in Gutenberg's palette of blues:

<img width="336" alt="Screenshot 2019-05-02 at 11 43 01" src="https://user-images.githubusercontent.com/376315/57099341-15219d80-6cd1-11e9-9a15-9ab15f184ca9.png">

https://github.com/WordPress/gutenberg/blob/c2c8276ac8e89af8382d5180ffa6a52f7fa3f98e/assets/stylesheets/_colors.scss

We might want to review how blues are being used in an attempt to scale down the number of different shades in use.

**Comment by afercia** [comment]
Ideally, the WordPress palette and the Gutenberg palette should be unified, in an effort to get the best of both worlds. I don't see great value in keeping two separate sets of colors, especially in terms of maintenance and consistency.

**Comment by sarahmonster** [comment]
Agreed! I'd really love to see our palette unified and pared down a wee bit.

**Comment by afercia** [comment]
Glad to hear there's consensus! That's definitely a job for designers. I'd expect from the design team to open a Trac ticket and lead the effort to unify the two palettes 🙂

**Comment by afercia** [comment]
Note for triage: only one part of this issue should be moved to core Trac. Please keep this issue open.

**Comment by afercia** [comment]
Note 2: the Publish button was fixed in #15544. The Calendar selected date still needs improvements.

**Comment by afercia** [comment]
As mentioned in https://github.com/WordPress/gutenberg/issues/15432#issuecomment-489544687 here's the relevant core Trac tickets:

**The design of the focus outline on buttons/elements could be improved**
https://core.trac.wordpress.org/ticket/34904
With some design proposals from @MichaelArestad 

**a11y-focus: Standardizing the handling of :focus and :hover**
https://core.trac.wordpress.org/ticket/34957

So maybe there's no need to move the first part of this issue to Trac, as there are already tickets about it.

**Comment by joyously** [comment]
This issue should be closed by removing all opinionated styles from the calendar block. (and all blocks!)
There is text color, background color, padding, width of 100%, link color.

Here is what the block styles produce on the front end:
![calendar-block-on-dark](https://user-images.githubusercontent.com/7957680/62841030-c341c980-bc68-11e9-80c5-5bafbba02fe6.jpg)

and here is how my theme styles it without any additional specific styling (I just removed the wp-block-calendar class using Dev Tools):
![theme-calendar-block-on-dark](https://user-images.githubusercontent.com/7957680/62841035-e8ced300-bc68-11e9-8b40-e8d5e3674d04.jpg)

**Comment by tellthemachines** [comment]
Current status of this ticket:

* Trac tickets already exist as per [@afercia's comment](https://github.com/WordPress/gutenberg/issues/15278#issuecomment-491277043)
* Publish button background has been changed to an accessible colour
* Calendar selected date background remains to be changed, but could possibly be addressed as part of #15962.

 As soon as that last item is done, it should be safe to close this issue, or have I missed anything?

**Comment by joedolson** [comment]
I haven't been able to trace exactly where it changes, but in looking at the current color contrast for the selected date, the color is now rendering as `#007cba`, which passes contrast tests at 4.57:1 (https://jdlsn.com/color/?type=hex&color=FFFFFF&color2=007cba). I think that means that this issue can now be closed.

**Comment by mapk** [comment]
I think you're right, @joedolson. Here's a screenshot of the current state.

![Screen Shot 2020-07-24 at 1 03 36 PM](https://user-images.githubusercontent.com/617986/88431091-7dfda580-cdae-11ea-83ec-39e570978621.png)

I'll close it out. Thanks everyone!
